packages/sdk/src/evm/core/classes/contract-deployer.ts

```diff
@@ -1034,4 +1091,257 @@ export class ContractDeployer extends RPCConnectionHandler {
       return constructorParamValues[index];
     });
   }
+
+  private async computeAndCacheInfraContractsData() {
+    const chainId = (await this.getProvider().getNetwork()).chainId;
+
+    for (const value of Object.values(INFRA_CONTRACTS_MAP)) {
+      this.computeAddressInfra(value.contractType, chainId);
+    }
+  }
+
+  async computeAddressInfra(contractType: InfraContractType, chainId: number) {
+    const contract = INFRA_CONTRACTS_MAP[contractType];
+    if (
+      !this.infraContractsInfoCache[chainId] ||
+      !this.infraContractsInfoCache[chainId][contract.contractType]
+    ) {
+      if (contract.contractType === NativeTokenWrapper.contractType) {
+        const address = await computeDeploymentAddress(
+          WETHBytecode,
+          [],
+          await this.computeCreate2FactoryAddress(),
+        );
+
+        this.infraContractsInfoCache[chainId] = {
+          ...this.infraContractsInfoCache[chainId],
+          nativeTokenWrapper: {
+            predictedAddress: address,
+            bytecode: "",
+            encodedArgs: "",
+          },
+        };
+      } else {
+        let uri = "";
+
+        if (this.infraContractURICache[contract.contractType]) {
+          uri = this.deployMetadataCache[contract.contractType];
+        } else {
+          // fetch the publish URI from the ContractPublisher contract
+          const publishedContract = await new ThirdwebSDK("polygon")
+            .getPublisher()
+            .getVersion(THIRDWEB_DEPLOYER, contract.name);
+          if (!publishedContract) {
+            throw new Error(
+              `No published contract found for ${contract.name} at version by '${THIRDWEB_DEPLOYER}'`,
+            );
+          }
+          uri = publishedContract?.metadataUri;
+        }
+
+        const metadata = await this.fetchAndCacheDeployMetadata(uri);
+        const encodedArgs = (
+          await this.encodeConstructorParamsForImplementation(
+            metadata.compilerMetadata,
+            chainId,
+          )
+        ).encodedArgs;
+        const address = await computeDeploymentAddress(
+          metadata.compilerMetadata.bytecode,
+          encodedArgs,
+          await this.computeCreate2FactoryAddress(),
+        );
+
+        if (this.infraContractsInfoCache[chainId]) {
+          this.infraContractsInfoCache[chainId][contract.contractType] = {
+            predictedAddress: address,
+            bytecode: metadata.compilerMetadata.bytecode,
+            encodedArgs: encodedArgs,
+          };
+        } else {
+          this.infraContractsInfoCache[chainId] = {
+            [contract.contractType]: {
+              predictedAddress: address,
+              bytecode: metadata.compilerMetadata.bytecode,
+              encodedArgs: encodedArgs,
```

Comment:
when is this cache actually gonna be hit? bytecode/encodedArgs is only useful for the very first deploy right?

it feels like we're over complicating the code quite a bit with this cache... and it doesnt seem that useful. What we want to avoid is having to recheck if the infra is there deployed right? why don't we just cache that?

like right now even with your cache, we still do the RPC call for `getCode` everytime, isnt that the most expensive thing to run?

I would first focus on getting the flow right, then we can optimize where it makes sense, this cache logic feels error prone to me